webapp/src/ts/services/user-replace.service.ts

```diff
@@ -0,0 +1,110 @@
+import { Injectable } from '@angular/core';
+import { DbService } from '@mm-services/db.service';
+import { DBSyncService, SyncStatus } from '@mm-services/db-sync.service';
+import { SessionService } from '@mm-services/session.service';
+import { SettingsService } from '@mm-services/settings.service';
+import { UserSettingsService } from '@mm-services/user-settings.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UserReplaceService {
+  constructor(
+    private settingsService: SettingsService,
+    private userSettingsService: UserSettingsService,
+    private dbService: DbService,
+    private dbSyncService: DBSyncService,
+    private sessionService: SessionService,
+  ) {
+    const that = this;
+    this.settingsService
+      .get()
+      .then((settings) => {
+        if (settings?.transitions?.user_replace) {
+          that.dbSyncService.subscribe(async(status) => {
+            return that.syncStatusChanged(status);
+          });
+        }
+      });
+  }
+
+  async getUserContact() {
+    const { contact_id }: any = await this.userSettingsService.get();
+    if (!contact_id) {
+      return;
+    }
+    return this.dbService
+      .get()
+      .get(contact_id)
+      .catch(err => {
+        if (err.status === 404) {
+          return;
+        }
+        throw err;
+      });
+  }
+
+  setReplaced(originalContact, newContact) {
+    if (!originalContact) {
+      throw new Error('The original contact could not be found when replacing the user.');
+    }
+    if (!newContact) {
+      throw new Error('The new contact could not be found when replacing the user.');
+    }
+    if (!originalContact.parent?._id || originalContact.parent._id !== newContact.parent._id) {
+      throw new Error('The new contact must have the same parent as the original contact when replacing a user.');
+    }
+    // TODO Currently the transitions are not run for online users so this is not used.
+    const status = this.sessionService.isOnlineOnly() ? ReplaceStatus.READY : ReplaceStatus.PENDING;
+    originalContact.replaced = { status, by: newContact._id };
+  }
+
+  isReplaced(contact): boolean {
+    return contact.replaced;
+  }
```

Comment:
Ah, TIL about the `!!` operator!  (Though I am still not sure I understand why I should or should not explicitly set the return type on a function.  What criteria do you use for that?)

Reply:
Typescript is smart enough to "guess" (or infer) primitive types and sometimes even objects without having to manually type them as long as there is no variable typed `any` (which basically means "you're on your own, I can't typecheck that").  
Since we're not using typescript's strict mode in our codebase, some variables are left with no type and typescript interprets them as `any` and we're left on our own to type things.

In this scenario, `contact` is not typed so typescript considers its type as `any`. If it were properly typed, you wouldn't have been able to type the return type as `boolean` because `replaced` is an object of the shape `{ status: string; by: string; }` and it has no overlap with `boolean`. But because it was typed as `any`, typescript let you cast it to `boolean`. It's one of typescript's escape hatch and there is nothing wrong using it when you want to tell typescript "let me handle this, I know what I'm doing". In reality `isReplaced(contact) { return contact.replaced }` returns an object and not a boolean so it can get a bit misleading :sweat_smile: 

With all that said, to answer your question, my criteria is: if typescript can infer my function's return type then I don't explicitly set the return type.  
But if for some reason typescript can't then I try to type things to make it understand what I'm doing. There are some cases where it's simply not worth it to type everything it needs to understand so I will explicitly set the return type.

I hope that answers your question haha

Side note: the `!!` operator is just the `!` operator used twice, you can use it as many times as you want but in the end `!!!!something` is the same as `!!something`

Reply:
Ahhh, that makes a lot of sense!  Thanks for the explanation!